File: src/ts/ssh/sshSessionAlgorithms.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Disposable } from 'vscode-jsonrpc';
import {
	Cipher,
	Signer,
	Verifier,
	MessageSigner,
	MessageVerifier,
	CompressionAlgorithm,
} from './algorithms/sshAlgorithms';

export class SshSessionAlgorithms implements Disposable {
	public publicKeyAlgorithmName?: string;
	public cipher?: Cipher | null;
	public decipher?: Cipher | null;
	public signer?: Signer | null;
	public verifier?: Verifier | null;
	public messageSigner?: MessageSigner | null;
	public messageVerifier?: MessageVerifier | null;
	public compressor?: CompressionAlgorithm | null;
	public decompressor?: CompressionAlgorithm | null;

	public dispose(): void {
		if (this.cipher) this.cipher.dispose();
		if (this.decipher) this.decipher.dispose();
		if (this.signer) this.signer.dispose();
		if (this.verifier) this.verifier.dispose();
	}
}

File: src/cs/Ssh/SshSessionAlgorithms.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using Microsoft.DevTunnels.Ssh.Algorithms;

namespace Microsoft.DevTunnels.Ssh;

/// <summary>
/// Algorithms that are negotiated between server and client for a session.
/// </summary>
internal class SshSessionAlgorithms : IDisposable
{
	public string? PublicKeyAlgorithmName { get; set; }

	public ICipher? Decipher { get; set; }

	public ICipher? Cipher { get; set; }

	public ISigner? Signer { get; set; }

	public IVerifier? Verifier { get; set; }

	public IMessageSigner? MessageSigner { get; set; }

	public IMessageVerifier? MessageVerifier { get; set; }

	public CompressionAlgorithm? Compressor { get; set; }

	public CompressionAlgorithm? Decompressor { get; set; }

	public void Dispose()
	{
		Dispose(true);
	}

	protected virtual void Dispose(bool disposing)
	{
		if (disposing)
		{
			Decipher?.Dispose();
			Cipher?.Dispose();
			Signer?.Dispose();
			Verifier?.Dispose();
			Compressor?.Dispose();
			Decompressor?.Dispose();
		}
	}
}
